app/server/utils.py

```diff
@@ -74,16 +76,80 @@ def exclude_created_labels(self, labels, created):
         """
         return [label for label in labels if label not in created]
 
-    def to_serializer_format(self, labels):
-        """Exclude created labels.
+    @classmethod
+    def to_serializer_format(cls, labels, created):
+        """Convert a label to model dictionary.
+
+        Also assigns shortkeys for each label that don't clash with existing
+        label shortkeys.
 
         Example:
             >>> labels = ["positive"]
-            >>> self.to_serializer_format(labels)
-            [{"text": "negative"}]
-        ```
+            >>> created = {}
+            >>> BaseStorage.to_serializer_format(labels, created)
+            [{"text": "positive", "suffix_key": "p", "prefix_key": None}]
         """
-        return [{'text': label} for label in labels]
+        existing_shortkeys = {(label.suffix_key, label.prefix_key)
+                              for label in created.values()}
+
+        serializer_labels = []
+
+        for label in sorted(labels):
+            serializer_label = {'text': label}
+
+            shortkey = cls.get_shortkey(label, existing_shortkeys)
+            if shortkey:
+                serializer_label['suffix_key'] = shortkey[0]
+                serializer_label['prefix_key'] = shortkey[1]
+                existing_shortkeys.add(shortkey)
+
+            background_color = cls.generate_color()
+            text_color = cls.black_or_white(background_color)
+            serializer_label['background_color'] = background_color
+            serializer_label['text_color'] = text_color
+
+            serializer_labels.append(serializer_label)
+
+        return serializer_labels
+
+    @classmethod
+    def get_shortkey(cls, label, existing_shortkeys):
+        """Find the first non existing shortkey for the label.
+
+        Example without existing shortkey:
+            >>> BaseStorage.get_shortkey("positive", set())
+            ("p", None)
+
+        Example with existing shortkey:
+            >>> BaseStorage.get_shortkey("positive", {("p", None)})
+            ("p", "ctrl")
+        """
+        model_prefix_keys = [key for (key, _) in Label.PREFIX_KEYS]
+        prefix_keys = [None] + model_prefix_keys
+
+        model_suffix_keys = {key for (key, _) in Label.SUFFIX_KEYS}
+        suffix_keys = [key for key in label.lower() if key in model_suffix_keys]
+
+        for shortkey in itertools.product(suffix_keys, prefix_keys):
+            if shortkey not in existing_shortkeys:
+                return shortkey
+
+        return None
+
+    @classmethod
+    def generate_color(cls):
```

Comment:
I think `generate_color` and `black_or_white` are implemented the same as one in JavaScript. But this may be better in a separate class or function like:

```python
import random


class Color(object):

    def __init__(self, rgb=None):
        self._rgb = rgb or random.sample(range(256), k=3)

    def contrast_color(self):
        """Generate black or white color.                                                                                                                              
                                                                                                                                                                       
        Ensure that foreground and background color combinations provide                                                                                               
        sufficient contrast when viewed by someone having color deficits or                                                                                            
        when viewed on a black and white screen.                                                                                                                       
                                                                                                                                                                       
        Algorithm from w3c:                                                                                                                                            
        * https://www.w3.org/TR/AERT/#color-contrast                                                                                                                   
        """
        r, g, b = self._rgb
        is_white = (((r * 299) + (g * 587) + (b * 114)) / 1000) < 128
        return Color([255] * 3) if is_white else Color([0] * 3)

    def to_hex(self):
        """Convert rgb to hex color."""
        return '#{:02x}{:02x}{:02x}'.format(*self._rgb)


if __name__ == '__main__':
    background_color = Color()
    text_color = background_color.contrast_color()
    print(background_color.to_hex())
    print(text_color.to_hex())
```

Reply:
Done in c996d10.